Don't allow global memberships on placeholder users

oliverguenther opened the pull request:
The view_all_principals migration added global roles to all users with manage_members.

On QA, we had placeholder users with this permission. Placeholder users should never receive global roles, but there was no validation in place.

This commit adds a migration to remove those, and adds a validation to the contract.
https://community.openproject.org/work_packages/68655

oliverguenther committed:
Don't allow global memberships on placeholder users

The view_all_principals migration added global roles to all users with
manage_members.

On QA, we had placeholder users with this permission. Placeholder users
should never receive global roles, but there was no validation in place.

This commit adds a migration to remove those, and adds a validation to
the contract.

cbliard commented on app/contracts/members/base_contract.rb:
```diff
@@ -50,6 +52,12 @@ def assignable_projects
 
     private
 
+    def validate_no_global_placeholder
+      return unless principal.is_a?(PlaceholderUser)
+
+      errors.add :principal, :invalid if project_id.nil?
+    end
+
```
Even without being a `PlaceholderUser`, validation would fail because `project_id` is blank.

Prying into the spec you added, the contract fails with the following errors: `{roles: ["has an unassignable role."], project: ["can't be blank."], principal: ["is invalid."]}` (I was curious to see the error message).

So it can't hurt, but I'm not sure that's actually useful.

cbliard replied:
But pushing it further, I actually see that this "project can't be blank" error is not added if current user is admin.
So only admins can set global roles?

So to make the unit test more valid, the current user should be an admin, right? so the other errors about roles and project do not trigger.

I'll continue reviewing tomorrow.

cbliard replied:
🟢 ok, It think it's fine to ignore the other errors reported for not being an admin as the test checks for another error anyway.

cbliard reviewed (commented):
I'm still reviewing. I'll continue tomorrow.

cbliard commented on app/models/member.rb:
```diff
@@ -152,4 +152,8 @@ def validate_presence_of_principal
+
+  def placeholder?
+    principal.is_a?(PlaceholderUser)
+  end
```
🟢 This method is not used, and it's private. It could be removed, or made public and used in base_contract but not sure if that's worth it.

oliverguenther replied:
Removed 👍

cbliard reviewed (approved):
Nice 👍

cbliard committed:
Fix flaky spec

failing spec: `spec/features/admin/enterprise/enterprise_spec.rb:130`
failing run: `https://github.com/opf/openproject/actions/runs/19163991005/job/54780241287?pr=20840`

Clicking on some elements in modals can fail as the click location is
wrong if the modal animation is not finished yet.

Fix the test by waiting for anmiation complete, and extract a page
object for enterprise tokens page to prevent repetition.
